include/matx/kernels/channelize_poly.cuh

```diff
@@ -0,0 +1,440 @@
+#include <memory>
+#include <stdint.h>
+#include <stdio.h>
+#include <vector>
+
+#include "matx/core/utils.h"
+#include "matx/core/type_utils.h"
+#include "matx/core/tensor_utils.h"
+
+namespace matx {
+
+// Number of output elements generated per thread
+constexpr index_t CHANNELIZE_POLY1D_ELEMS_PER_THREAD = 1;
+
+#ifdef __CUDACC__ 
+template <int THREADS, typename OutType, typename InType, typename FilterType>
+__launch_bounds__(THREADS)
+__global__ void ChannelizePoly1D(OutType output, InType input, FilterType filter)
+{
+    using output_t = typename OutType::scalar_type;
+    using input_t = typename InType::scalar_type;
+    using filter_t = typename FilterType::scalar_type;
+
+    // Opportunistically store the filter taps in shared memory if the static shared memory
+    // size is sufficient. Otherwise, we will read directly from global memory on use.
+    const int SMEM_MAX_FILTER_TAPS = 128;
+    __shared__ filter_t smem_filter[SMEM_MAX_FILTER_TAPS];
+
+    constexpr int InRank = InType::Rank();
+    constexpr int OutRank = OutType::Rank();
+    constexpr int ChannelRank = OutRank-1;
+    constexpr int OutElemRank = OutRank-2;
+
+    const index_t input_len = input.Size(InRank-1);
+    const index_t output_len_per_channel = output.Size(OutElemRank);
+    const index_t num_channels = output.Size(ChannelRank);
+    const index_t filter_full_len = filter.Size(0);
+    const index_t filter_phase_len = (filter_full_len + num_channels - 1) / num_channels;
+
+    const int elem_block = blockIdx.x;
+    const int channel = blockIdx.y;
+    const int tid = threadIdx.x;
+
+    constexpr index_t ELEMS_PER_BLOCK = CHANNELIZE_POLY1D_ELEMS_PER_THREAD * THREADS;
+    const index_t first_out_elem = elem_block * CHANNELIZE_POLY1D_ELEMS_PER_THREAD * THREADS;
+    const index_t last_out_elem = std::min(
+        output_len_per_channel - 1, first_out_elem + ELEMS_PER_BLOCK - 1);
+
+    if (filter_phase_len <= SMEM_MAX_FILTER_TAPS) {
+        for (index_t t = tid; t < filter_phase_len-1; t += THREADS) {
+            const index_t h_ind = channel + t * num_channels;
+            smem_filter[t] = filter.operator()(h_ind);
+        }
+        if (tid == THREADS-1) {
+            const index_t h_ind = channel + (filter_phase_len-1) * num_channels;
+            smem_filter[filter_phase_len-1] = (h_ind < filter_full_len) ?
+                filter.operator()(h_ind) : static_cast<filter_t>(0);
+        }
+
+        __syncthreads();
+    }
+
+    auto indims = BlockToIdx(input, blockIdx.z, 1);
+    auto outdims = BlockToIdx(output, blockIdx.z, 2);
+    outdims[ChannelRank] = channel;
+
+    // For C2C transforms, we use ifft() to get the correct sign on the imaginary
+    // components and thus avoid a post-fft conjugation. However, ifft scales by
+    // 1/N, so we pre-scale here by N to compensate. For R2C transforms, we must
+    // use fft() and we handle conjugation in the subsequent unpacking.
+    const auto scale = [num_channels]() -> auto {

```

Comment:
Jonathan changed this recently to add different types of normalization. It seems maybe you want a new type of "none" so you don't have IFFT scale at all?

Reply:
Yes, if we can avoid scaling after the ifft, then this scaling can be removed. It looks like we can accomplish that now by setting the FFTNorm mode to FORWARD in the ifft, although NONE may better capture the intent. I can update the PR to do either one (add a NONE scaling mode, or just specify forward for the ifft).

Reply:
Updated to use FFTNorm::FORWARD with ifft and remove the scaling from the kernel.